UI Mixed State changes

designedbyscience opened the pull request:
Here are some updates for Mixed States.

Additionally, some small layout adjustments and design changes.

Notably, the font preview swatch no longer reflect the size of the font

Mixed Vector Layers:
![untitled-1___51_6___rgb_8____](https://cloud.githubusercontent.com/assets/22372/9798526/7791aa02-57cf-11e5-8d8d-e5bdd1b1f32b.png)

Mixed Vector and Type:
![untitled-1___51_6___rgb_8____](https://cloud.githubusercontent.com/assets/22372/9798219/99647eae-57cd-11e5-8765-dfb5e1f76263.png)

Mixed Type, same color, different faces
![untitled-1___51_6___rgb_8____](https://cloud.githubusercontent.com/assets/22372/9798237/b695e76a-57cd-11e5-99cc-e8d264d67d19.png)

Mixed Type Layers
![untitled-1___51_6___rgb_8____](https://cloud.githubusercontent.com/assets/22372/9798582/e0a61a32-57cf-11e5-9f14-2c2ef96bb08e.png)

shaoshing commented:
Hey @designedbyscience, you may already noticed, but here are some issues I encountered when testing the mix state:

1. I think we should remove these attributes for empty layers and art boards? (and they are not working)
 <img width="450" alt="screen shot 2015-09-10 at 2 49 03 pm" src="https://cloud.githubusercontent.com/assets/118264/9802661/e5628c1c-57d0-11e5-8e2c-696f5628c54b.png">

2. Shall we show text preview (and text color) for mixed text and shape layers? I am asking this because I found myself looking for text button and realized it got merged into the stroke color button. 
<img width="783" alt="screen shot 2015-09-10 at 2 55 52 pm" src="https://cloud.githubusercontent.com/assets/118264/9802877/6f501c86-57d2-11e5-8804-1155921150bd.png">
 <img width="450" alt="screen shot 2015-09-10 at 2 55 58 pm" src="https://cloud.githubusercontent.com/assets/118264/9802733/78f7adc2-57d1-11e5-8611-c6e7d7bec142.png">

3. How about mixed font style?
 <img width="450" alt="screen shot 2015-09-10 at 3 04 06 pm" src="https://cloud.githubusercontent.com/assets/118264/9802763/9d25c706-57d1-11e5-90ba-3183c1b899c1.png">

4. Stroke color and border attributes should not shown as mixed, if I select an empty layer and a shape layer.
<img width="769" alt="screen shot 2015-09-10 at 3 20 11 pm" src="https://cloud.githubusercontent.com/assets/118264/9802810/e07c7856-57d1-11e5-9380-03a4c0f5ff58.png">

5. I am unable to set stroke and border position (maybe there are more) when select text and shape layers
![ui-mix-update](https://cloud.githubusercontent.com/assets/118264/9802844/1c2545e0-57d2-11e5-9c7b-09f041ef0ef6.gif)

designedbyscience commented:
In general, I think these questions bring up some questions about consistency in behavior which I'll discuss with @placegraphichere (some might be bugs, some might be things we want to change)

iwehrman commented:
Also note that there may be some actions that didn't work correctly previously with mixed selections. In that case, we should file bugs about those issues instead of blocking this PR. Someone other than @designedbyscience should be tossed onto those non-UI issues.

designedbyscience commented:
Okay...we changed a bunch of things in this.

Big change/rule to be aware of:

For mixed vector/type selections, we are disabling any control that doesn't apply to all
Disabled inputs appear with NO content

Mixed but not disabled inputs should have -- in them

@shaoshing back to you

shaoshing commented on src/js/jsx/sections/style/AppearancePanel.jsx:
```diff
@@ -152,13 +154,23 @@ define(function (require, exports, module) {
-            
+
+            var selectedLayers = this.props.document.layers.selected,
+                mixedVectorType = selectedLayers.some(function (layer) {
```
`mixedVectorType` does not sound like a boolean. Maybe call it `hasMixedVectorType`?

shaoshing commented on src/js/jsx/sections/style/Stroke.jsx:
```diff
@@ -272,18 +288,11 @@ define(function (require, exports, module) {
-                            onClick={!this.props.disabled ? this._toggleStrokeEnabled : _.noop}
-                            size="column-2"
-                        />
+                            {toggleVisibilityButton}
```
This line has extra indent :beetle:

shaoshing commented on src/js/jsx/sections/style/Type.jsx:
```diff
@@ -506,26 +517,53 @@ define(function (require, exports, module) {
+                            style={fillStyle}/>
+                    );
+                } else {
+                    typeStyle = {
```
Looks like `typeStyle` is only used in this scope and its initial value will always get overwritten. Maybe consider remove the outer `typeStyle` assignment?

shaoshing commented on src/js/jsx/sections/style/Type.jsx:
```diff
@@ -506,26 +517,53 @@ define(function (require, exports, module) {
+            
+            if (this.props.mixedVectorType) {
+                fillBlendFormline = null;
+            } else {
```
Assigning `fillBlendFormline` to null seems not necessary here, as the default value `undefined` will do the job.

shaoshing commented on src/js/jsx/sections/style/Type.jsx:
```diff
@@ -597,31 +641,31 @@ define(function (require, exports, module) {
                                     onClick={this._handleAlignmentChange.bind(this, textLayer.alignmentTypes.RIGHT)}
                                     title={strings.TOOLTIPS.ALIGN_TYPE_RIGHT} />
                                 <SplitButtonItem
+                                    disabled={locked || !box}
```
`box` is only used here but is defined at line 455 (200 lines away). can you move it closer to where it is used?

designedbyscience replied:
Gonna push back on this one: I think that grouping `box` with the other collection operations makes more sense than moving it someplace closer

shaoshing commented on src/js/jsx/sections/style/VectorAppearance.jsx:
```diff
@@ -44,19 +43,16 @@ define(function (require, exports, module) {
-            });
-
-            if (onlyTextLayers) {
+            if (this.props.uniformLayerKind === this.props.document.layers.byIndex(0).layerKinds.TEXT) {
```
Do you think we can import these constants from the adapter and refer to TEXT directly, instead of reading it from layer instance? For example:

``` 
var LAYER_KINDS = require("adapter/lib/layer").layerKinds
if (this.props.uniformLayerKind === LAYER_KINDS.TEXT) {
```

Or expose them to the Layer class 

```
var Layer = require("js/models/layer");
if (this.props.uniformLayerKind === Layer.KINDS.TEXT) {
```

designedbyscience replied:
This seems like a reasonable idea, however, I think you should file that as a separate code cleanup task (as, the pattern I used here is the typical pattern for JSX throughout the rest of the code)

shaoshing commented on src/js/jsx/sections/style/VectorFill.jsx:
```diff
@@ -111,20 +111,22 @@ define(function (require, exports, module) {
                 enabledFlags: enabledFlags
             };
         },
-
```
line break please :bee:

shaoshing commented on src/js/jsx/sections/style/VectorFill.jsx:
```diff
@@ -111,20 +111,22 @@ define(function (require, exports, module) {
-            });
-
-            if (onlyTextLayers) {
+            if (this.props.uniformLayerKind === this.props.document.layers.byIndex(0).layerKinds.TEXT) {
```
same here for the layer kind constants.

shaoshing commented:
Should the leading and tracking icons be greyout as well?
<img width="741" alt="screen shot 2015-09-18 at 4 06 29 pm" src="https://cloud.githubusercontent.com/assets/118264/9973367/900ae67c-5e26-11e5-85d7-0d54cb9225ad.png">

These don't seem right for mixed text & smart object, and text & artboard
<img width="768" alt="screen shot 2015-09-18 at 4 08 16 pm" src="https://cloud.githubusercontent.com/assets/118264/9973376/b9b7dce6-5e26-11e5-8a47-91e6989237de.png">
<img width="708" alt="screen shot 2015-09-18 at 4 09 37 pm" src="https://cloud.githubusercontent.com/assets/118264/9973379/bec2699a-5e26-11e5-9579-9e91874475db.png">

shaoshing commented:
Done reviewing! I have couple nits about the code, and two questions about the mix state, otherwise looks awesome! Back to you @designedbyscience

designedbyscience commented:
Rewrote a lot of mixed state detection code, seems cleaner now (more information passed through props, less looking at the layer set inside of components)

Also, addressed the visual quirks above

shaoshing commented:
Bravo :beers:
